src/sinks/clickhouse/native_sink.rs

```diff
@@ -0,0 +1,84 @@
+use std::collections::BTreeMap;
+
+use futures::{stream::BoxStream, StreamExt};
+use clickhouse_rs::{Pool, types::SqlType,};
+use vector_core::stream::BatcherSettings;
+use async_trait::async_trait;
+
+use super::{ClickhouseConfig, native_service::ClickhouseService, parse::parse_field_type};
+use crate::{
+    config::{SinkContext, SinkHealthcheckOptions},
+    event::Event,
+    sinks::{
+        util::{StreamSink, SinkBuilderExt},
+        Healthcheck, VectorSink,
+    },
+};
+
+pub(super) async fn build_native_sink(cfg: &ClickhouseConfig, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+    let table_schema = gen_table_schema(&cfg.table_def)?;
+    let batch = cfg.batch.into_batcher_settings()?;
+    let pool = Pool::new(cfg.endpoint.to_string());
+    let sink = NativeClickhouseSink::new(pool.clone(), batch, cfg.table.clone(), table_schema);
+    let health_check = healthcheck(pool, cx.healthcheck.clone());
+    Ok((
+        VectorSink::from_event_streamsink(sink),
+        Box::pin(health_check)
+    ))
+}
+
+fn gen_table_schema(table: &BTreeMap<String, String>) -> crate::Result<Vec<(String, SqlType)>> {
+    let mut table_schema = Vec::with_capacity(table.len());
+    for (k,v) in table {
+        match parse_field_type(v.as_str()) {
+            Ok((_, t)) => {
+                table_schema.push((k.to_owned(), t));
+            },
+            Err(e) => {
+                let ne: crate::Error = std::convert::From::from(e.to_string());
+                Err(ne)?;
+            }
+        }
+    }
+    Ok(table_schema)
```

Comment:
Could use a `map` and let Rust's type inference do some magic:
```suggestion
    table.into_iter().map(|(k, v)| {
        parse_field_type(v.as_str())
            .map(|(_, t)| (k.to_owned(), t))
            .map_err(|e| e.to_string().into())
    }).collect()
```